Issue 738: support empty publicID

AnjoMan opened the pull request:
This adds support for specifying a publicID as '';
instead of evaluating the falsy-ness of the input,
we want to use the provided string, even if it is
''.

Fixes #738

## Proposed Changes

  - when the user specifies `publicID` as '', we should not trigger the `None` default behaviour

AnjoMan commented:
If you like this contribution, please add the #hacktoberfest-accepted label

ashleysommer commented:
Hi @AnjoMan
This looks like a good change, we will discuss this more formally tomorrow at our RDFLib maintainers meeting. We will have to run some more tests internally to ensure everything in the RDF/XML serializer operates correctly after this change.

I'm sorry, we are not participating in DigitalOcean's Hacktoberfest campaign.

ashleysommer commented:
Hi @AnjoMan 
After discussing this, we have decided we will need to test this solution further. You have implemented the change just for the RDF/XML parser, but we need to test what happens if you use `publicID=""` for other parsers, like TRIG, or N3, or JSON-LD. While your change does fix the problem you are seeing, it could also potentially introduce breakages in unexpected places, so we need to be cautious about how to proceed with this.

white-gecko reviewed (commented):
As can be seen for the turtle example, with your fix we would introduce some inconsistency across the serializers. The fix does also not work in the same way for the current Turtle parser/notation3 parser as we call `graph.absolutize()` in this case.

What is the usecase for an emptyID?

In m eyes this relative URI thing without explicitly defining a base in the file is a tough problem.

white-gecko reviewed (changes requested):
This issue needs to be solved for all parsers at once.

AnjoMan committed:
Issue 738: support empty publicID

This adds support for specifying a publicID as '';
instead of evaluating the falsy-ness of the input,
we want to use the provided string, even if it is
''.

AnjoMan committed:
Try to fix failing notation3 tests

AnjoMan committed:
Add tests of empty publicID for all parsers

AnjoMan committed:
empty publicID supported by rdfxml parser

AnjoMan committed:
empty publicID supported by notation3 parser

AnjoMan committed:
empty publicID supported by trig parser

AnjoMan committed:
empty publicID supported by ntriples / nquads parsers

AnjoMan committed:
empty publicID supported by json-ld parser

AnjoMan committed:
Refactor tests of json-ld norm_url function
